File: cpp_editor/src/c/HabaParser.cpp
#include "HabaParser.h"
#include <regex>
#include <sstream>
#include <vector>

namespace {
// Helper function to trim leading/trailing whitespace from a string
std::string trim(const std::string& str) {
    const std::string whitespace = " \t\n\r\f\v";
    size_t first = str.find_first_not_of(whitespace);
    if (std::string::npos == first) {
        return str;
    }
    size_t last = str.find_last_not_of(whitespace);
    return str.substr(first, (last - first + 1));
}

// Helper function to split a string by a delimiter
std::vector<std::string> split(const std::string& str, char delimiter) {
    std::vector<std::string> tokens;
    std::string token;
    std::istringstream tokenStream(str);
    while (std::getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}
}

HabaData HabaParser::parse(const std::string& raw_text) {
    HabaData data;
    std::smatch match;

    // Extract content layer
    std::regex content_regex("<content_layer>(.*?)</content_layer>", std::regex::ECMAScript | std::regex::multiline);
    if (std::regex_search(raw_text, match, content_regex) && match.size() > 1) {
        data.content = trim(match[1].str());
    }

    // Extract presentation layer
    std::regex presentation_regex("<presentation_layer>(.*?)</presentation_layer>", std::regex::ECMAScript | std::regex::multiline);
    if (std::regex_search(raw_text, match, presentation_regex) && match.size() > 1) {
        std::string presentation_text = match[1].str();

        // Extract containers
        std::smatch containers_match;
        std::regex containers_regex("<containers>(.*?)</containers>", std::regex::ECMAScript | std::regex::multiline);
        std::vector<std::string> containers;
        if (std::regex_search(presentation_text, containers_match, containers_regex) && containers_match.size() > 1) {
            std::string containers_block = trim(containers_match[1].str());
            auto lines = split(containers_block, '\n');
            for(const auto& line : lines) {
                std::string trimmed_line = trim(line);
                if (!trimmed_line.empty()) {
                    containers.push_back(trimmed_line);
                }
            }
        }

        // Extract styles
        std::smatch styles_match;
        std::regex styles_regex("<styles>(.*?)</styles>", std::regex::ECMAScript | std::regex::multiline);
        std::vector<std::string> styles;
        if (std::regex_search(presentation_text, styles_match, styles_regex) && styles_match.size() > 1) {
            std::string styles_block = trim(styles_match[1].str());
            auto lines = split(styles_block, '\n');
            for(const auto& line : lines) {
                std::string trimmed_line = trim(line);
                if (!trimmed_line.empty()) {
                    styles.push_back(trimmed_line);
                }
            }
        }

        // Match containers and styles by order
        for (size_t i = 0; i < containers.size(); ++i) {
            std::string style = (i < styles.size()) ? styles[i] : "";
            data.presentation_items.emplace_back(containers[i], style);
        }
    }

    // Extract script layer
    std::regex script_regex("<script_layer>(.*?)</script_layer>", std::regex::ECMAScript | std::regex::multiline);
    if (std::regex_search(raw_text, match, script_regex) && match.size() > 1) {
        data.script = trim(match[1].str());
    }

    return data;
}

std::string HabaParser::build(const HabaData& haba_data) {
    std::stringstream ss;

    // Build content layer
    ss << "<content_layer>\n    " << haba_data.content << "\n</content_layer>\n";

    // Build presentation layer
    ss << "<presentation_layer>\n";
    ss << "    <containers>\n";
    for (const auto& item : haba_data.presentation_items) {
        ss << "        " << item.first << "\n";
    }
    ss << "    </containers>\n";
    ss << "    <styles>\n";
    for (const auto& item : haba_data.presentation_items) {
        ss << "        " << item.second << "\n";
    }
    ss << "    </styles>\n";
    ss << "</presentation_layer>";

    // Build script layer
    if (!haba_data.script.empty()) {
        ss << "\n<script_layer>\n    " << haba_data.script << "\n</script_layer>";
    }

    return ss.str();
}

File: cpp_editor/reference/haba_parser.py
import re

class HabaData:
    """A simple data class to hold the parsed Haba file content."""
    def __init__(self):
        self.content = ""
        self.presentation_items = [] # A list of tuples (container_text, style_text)

class HabaParser:
    """
    A parser for the .haba file format.

    The format is described as an XML-like file where indentation is significant.
    It has a content_layer and a presentation_layer.
    In the presentation_layer, containers and styles are matched by order.
    """

    def parse(self, raw_text: str) -> HabaData:
        """
        Parses the raw text of a .haba file into a HabaData object.
        """
        data = HabaData()

        # Extract content layer
        content_match = re.search(r'<content_layer>(.*?)</content_layer>', raw_text, re.DOTALL)
        if content_match:
            data.content = content_match.group(1).strip()

        # Extract presentation layer
        presentation_match = re.search(r'<presentation_layer>(.*?)</presentation_layer>', raw_text, re.DOTALL)
        if presentation_match:
            presentation_text = presentation_match.group(1).strip()

            # Extract containers and styles
            containers_match = re.search(r'<containers>(.*?)</containers>', presentation_text, re.DOTALL)
            styles_match = re.search(r'<styles>(.*?)</styles>', presentation_text, re.DOTALL)

            containers = []
            if containers_match:
                # Assuming one container per line for now
                containers = [line.strip() for line in containers_match.group(1).strip().split('\n') if line.strip()]

            styles = []
            if styles_match:
                # Assuming one style per line for now
                styles = [line.strip() for line in styles_match.group(1).strip().split('\n') if line.strip()]

            # Match containers and styles by order
            for i in range(len(containers)):
                style = styles[i] if i < len(styles) else "" # Default to empty style if not enough styles
                data.presentation_items.append((containers[i], style))

        return data

    def build(self, haba_data: HabaData) -> str:
        """
        Builds a .haba file string from a HabaData object.
        """
        # Build content layer
        content_str = f"<content_layer>\n    {haba_data.content}\n</content_layer>\n"

        # Build presentation layer
        containers_str = "\n".join([f"        {item[0]}" for item in haba_data.presentation_items])
        styles_str = "\n".join([f"        {item[1]}" for item in haba_data.presentation_items])

        presentation_str = (
            "<presentation_layer>\n"
            "    <containers>\n"
            f"{containers_str}\n"
            "    </containers>\n"
            "    <styles>\n"
            f"{styles_str}\n"
            "    </styles>\n"
            "</presentation_layer>"
        )

        return content_str + presentation_str


# Example Usage (for testing purposes)
if __name__ == '__main__':
    example_haba_text = """
    <content_layer>
        This is the main text content.
        It can span multiple lines.
    </content_layer>
    <presentation_layer>
        <containers>
            <div>
            <p>
        </containers>
        <styles>
            { color: 'blue' }
            { font-size: '16px' }
        </styles>
    </presentation_layer>
    """

    parser = HabaParser()
    parsed_data = parser.parse(example_haba_text)

    print("--- PARSED DATA ---")
    print(f"Content: {parsed_data.content}")
    for i, (container, style) in enumerate(parsed_data.presentation_items):
        print(f"Item {i+1}: Container='{container}', Style='{style}'")

    print("\n--- REBUILT FILE ---")
    rebuilt_text = parser.build(parsed_data)
    print(rebuilt_text)
